babel/numbers.py

```diff
@@ -385,6 +384,205 @@ def format_scientific(number, format=None, locale=LC_NUMERIC):
+
+    >>> _find_unit_pattern("radian", locale="en")
+    'angle-radian:long'
+
+    Unknown values will return None.
+
+    >>> _find_unit_pattern("horse", locale="en")
+
+    :param unit_id: the code of a measurement unit.
+    :param length: The length for the formatting patterns (long, short or narrow)
+    :return: A key to the `unit_patterns` mapping, or None.
+    """
+    locale = Locale.parse(locale)
+    pattern = "%s:%s" % (unit_id.split(":", 1)[0], length)
+    if pattern in locale.unit_patterns:
+        return pattern
+    for unit_pattern in sorted(locale.unit_patterns, key=len):
+        if unit_pattern.endswith(pattern):
+            return unit_pattern
+
+
+def format_unit(value, measurement_unit, length='long', format=None, locale=LC_NUMERIC):
+    """Format a value of a given unit.
+
+    Values are formatted according to the locale's usual pluralization rules
+    and number formats.
+
+    >>> format_unit(12, 'length-meter', locale='ro_RO')
+    u'12 metri'
+    >>> format_unit(15.5, 'length-mile', locale='fi_FI')
+    u'15,5 mailia'
+    >>> format_unit(1200, 'pressure-inch-hg', locale='nb')
+    u'1\\xa0200 tommer kvikks\\xf8lv'
+
+    Number formats may be overridden with the ``format`` parameter.
+
+    >>> format_unit(Decimal("-42.774"), 'temperature-celsius', 'short', format='#.0', locale='fr')
+    u'-42,8 \\xb0C'
+
+    The locale's usual pluralization rules are respected.
+
+    >>> format_unit(1, 'length-meter', locale='ro_RO')
+    u'1 metru'
+    >>> format_unit(0, 'length-picometer', locale='cy')
+    u'0 picometr'
+    >>> format_unit(2, 'length-picometer', locale='cy')
+    u'2 bicometr'
+    >>> format_unit(3, 'length-picometer', locale='cy')
+    u'3 phicometr'
+
+    >>> format_unit(15, 'length-horse', locale='fi')
+    Traceback (most recent call last):
+        ...
+    ValueError: length-horse/long is not a known unit/length in fi
+
+    .. versionadded:: 2.2.0
+
+    :param value: the value to format.
+    :param measurement_unit: the code of a measurement unit.
+                             Known units can be found in the CLDR Unit Validity XML file:
+                             http://unicode.org/repos/cldr/tags/latest/common/validity/unit.xml
+    :param length: "short", "long" or "narrow"
+    :param format: An optional format, as accepted by `format_decimal`.
+    :param locale: the `Locale` object or locale identifier
+    """
+    locale = Locale.parse(locale)
+
+    unit_length = _find_unit_pattern(measurement_unit, length, locale)
+    if not unit_length:
+        raise ValueError("%s is not a known unit with length %s in %s" % (measurement_unit, length, locale))
+    unit_patterns = locale.unit_patterns[unit_length]
+    formatted_value = format_decimal(value, format, locale)
+    plural_form = locale.plural_form(value)
+    if plural_form in unit_patterns:
+        return unit_patterns[plural_form].format(formatted_value)
+    # Fall back to a bad representation.
```

Comment:
Would it be better to fall back to *any* plural form? eg. if 'many' doesn't exist, fall back to 'one' or 'zero' or 'few' or whatever is there.

If not, I think `measurement_unit` is pretty user unreadable. I'd prefer to fall back to `displayName` (which babel doesn't support) or show no unit at all.